src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -16,168 +17,383 @@
+
+		if (texturesInfo.setIx == samplersInfo.setIx)
+			layouts[texturesInfo.setIx] = utilities->getLogicalDevice()->createDescriptorSetLayout({ {textureBinding, samplersBinding} });
+		else
+		{
+			layouts[texturesInfo.setIx] = utilities->getLogicalDevice()->createDescriptorSetLayout({ {textureBinding} });
+			layouts[samplersInfo.setIx] = utilities->getLogicalDevice()->createDescriptorSetLayout({ {samplersBinding} });
+		}
+
+		assert(layouts[texturesInfo.setIx]);
+		assert(layouts[samplersInfo.setIx]);
+
+		return utilities->getLogicalDevice()->createPipelineLayout(PushConstantRanges, std::move(layouts[0u]), std::move(layouts[1u]), std::move(layouts[2u]), std::move(layouts[3u]));
+	}
+
+	const smart_refctd_ptr<IFileArchive> UI::mount(smart_refctd_ptr<ILogger> logger, ISystem* system, const std::string_view archiveAlias)
+	{
+		assert(system);
+		
+		if(!system)
+			return nullptr;
+
+		auto archive = make_smart_refctd_ptr<builtin::CArchive>(smart_refctd_ptr(logger));
+		system->mount(smart_refctd_ptr(archive), archiveAlias.data());
+
+		return smart_refctd_ptr(archive);
+	}
+
+	void UI::createPipeline(SCreationParameters& creationParams)
+	{
+		auto pipelineLayout = smart_refctd_ptr<IGPUPipelineLayout>(creationParams.pipelineLayout);
 
 		if (!pipelineLayout)
 		{
-			m_creationParams.utilities->getLogger()->log("Could not create pipeline layout!", system::ILogger::ELL_ERROR);
+			creationParams.utilities->getLogger()->log("Could not create pipeline layout!", ILogger::ELL_ERROR);
 			assert(false);
 		}
 
 		struct
 		{
-			core::smart_refctd_ptr<video::IGPUShader> vertex, fragment;
+			smart_refctd_ptr<IGPUShader> vertex, fragment;
 		} shaders;
 
 		{
 			constexpr std::string_view NBL_ARCHIVE_ALIAS = "nbl/ext/imgui/shaders";
 				
-			auto system = smart_refctd_ptr<system::ISystem>(m_creationParams.assetManager->getSystem());																	//! proxy the system, we will touch it gently
-			auto archive = make_smart_refctd_ptr<nbl::ext::imgui::builtin::CArchive>(smart_refctd_ptr<system::ILogger>(m_creationParams.utilities->getLogger()));		//! we should never assume user will mount our internal archive since its the extension and not user's job to do it, hence we mount only to compile our extension sources then unmount the archive
-			auto compiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(system));																				//! note we are out of default logical device's compiler set scope so also a few special steps are required to compile our extension shaders to SPIRV
+			//! proxy the system, we will touch it gently
+			auto system = smart_refctd_ptr<ISystem>(creationParams.assetManager->getSystem());
+
+			//! note we are out of default logical device's compiler set scope so also a few special steps are required to compile our extension shaders to SPIRV
+			auto compiler = make_smart_refctd_ptr<CHLSLCompiler>(smart_refctd_ptr(system));	
```

Comment:
there's a default compiler set in both ILogicalDevice and IAssetManager IIRC (or at least in ILogicalDevice)

Reply:
> or at least in ILogicalDevice

it just so happens that you won't use it because it is protected

> and IAssetManager

but here yes you can, updated & resolved

Reply:
what I'm saying is that you don't need to do 
```cpp
		auto compiler = set->getShaderCompiler(IShader::E_CONTENT_TYPE::ECT_HLSL);
``
to extract a compiler.

The correct compiler will be invoked when you do `ILogicalDevice::createShader` because the input `ICPUShader` has content type HLSL